File: BinarySearch/278.First-bad-version/accepted.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        left = 1
        right = n - 1

        while left <= right:
            mid = (left + right) // 2
            print("left:{}, right:{}, mid:{}".format(left,right,mid))
            if not isBadVersion(mid) and isBadVersion(mid+1):
                return mid + 1
            elif isBadVersion(mid):
                right = mid - 1
            else:
                left = mid + 1
        
        return 1

File: BinarySearch/744.find-smallest-letter-greater-than-target/accepted.cpp
class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        int left = 0;
        int right = letters.size() - 1;

        while(left <= right){
            int mid = left + (right - left) / 2;
            if(left == right && right == letters.size() - 1){
                break;
            }

            if(letters[mid] <= target && target < letters[mid + 1]){
                return letters[mid + 1];
            }else if (letters[mid] <= target){
                left = mid + 1;
            }else{
                right = mid - 1;
            }
        }
        return letters[0];

        
    }
};